docs/typescript.md

```diff
@@ -0,0 +1,315 @@
+  >
+    ( f: StateCreator<T, Mps, Mcs>
+    , name?: string
+    ) =>
+      StateCreator<T, Mps, Mcs>
+
+type LoggerImpl =
+  <T extends State>
+    ( f: PopArgument<StateCreator<T, [], []>>
+    , name?: string
+    ) =>
+      PopArgument<StateCreator<T, [], []>>
+
+const loggerImpl: LoggerImpl = (f, name) => (set, get, store) => {
+  type T = ReturnType<typeof f>
+  const loggedSet: typeof set = (...a) => {
+    set(...a)
+    console.log(...(name ? [`${name}:`] : []), get())
+  }
+  store.setState = loggedState
+
+  return f(loggedSet, get, store)
+}
+
+export const logger = loggerImpl as unknown as Foo
+
+type PopArgument<T extends (...a: never[]) => unknown> =
+  T extends (...a: [...infer A, infer _]) => infer R
+    ? (...a: A) => R
+    : never
+
+// ---
+
+const useStore = create<BearState>()(logger((set) => ({
+  bears: 0,
+  increase: (by) => set((state) => ({ bears: state.bears + by })),
+}), "bear-store"))
+```
+
+### Middleware that changes the store type
+
+```js
+import create, { State, StateCreator, StoreMutatorIdentifier, Mutate, StoreApi } from "zustand"
+
+type Foo =
+  < T extends State
+  , A
+  , Mps extends [StoreMutatorIdentifier, unknown][] = []
+  , Mcs extends [StoreMutatorIdentifier, unknown][] = []
+  >
+    ( f: StateCreator<T, [...Mps, ['foo', A]], Mcs>
+    , bar: A
+    ) =>
+      StateCreator<T, Mps, [['foo', A], ...Mcs]>
+
+declare module 'zustand' {
+  interface StoreMutators<S, A> {
+    foo: Write<Cast<S, object> { foo: A }>
+  }
+}
+
+type FooImpl =
+  <T extends State, A>
+    ( f: PopArgument<StateCreator<T, [], []>>
+    , bar: A
+    ) => PopArgument<StateCreator<T, [], []>>
+
+const fooImpl: FooImpl = (f, bar) => (set, get, _store) => {
+  type T = ReturnType<typeof f>
+  type A = typeof bar
+
+  const store = _store as Mutate<StoreApi<T>, [['foo', A]]>
+  store.foo = bar
+  return f(set, get, _store)
+}
+
+export const foo = fooImpl as unknown as Foo
+
+type PopArgument<T extends (...a: never[]) => unknown> =
+  T extends (...a: [...infer A, infer _]) => infer R
+    ? (...a: A) => R
+    : never
+
+type Write<T extends object, U extends object> =
+  Omit<T, keyof U> & U
+
+type Cast<T, U> =
+  T extends U ? T : U;
+
+// ---
+
+const useStore = create(foo(() => ({ bears: 0 }), "hello"))
+console.log(store.foo.toUpperCase())
+```
+
+### `create` without curried workaround
+
+The recommended way to use `create` is using the curried workaround ie `create<T>()(...)` because this enabled you to infer the store type. But for some reason if you don't want to use the workaround then you can pass the type parameters like the following. Note that in some cases this acts as an assertion instead of annotation, so it's not recommended.
+
+```ts
+import create, { State, StateCreator, StoreMutatorIdentifier, Mutate, StoreApi } from "zustand"
+
+interface BearState {
+  bears: number
+  increase: (by: number) => void
+}
+
+const useStore = create<
+  BearState,
+  [],
```

Comment:
Reading types in `react.ts` again, don't we only have two type arguments?

Reply:
Ah right